Reporting/ad-extensions-reporter.js

```diff
@@ -0,0 +1,395 @@
+    35: price,
+    38: promotion
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+// Functions
+
+function main() {
+    var spreadsheet = checkSpreadsheet(spreadsheetUrl, "Output Spreadsheet");
+    var sheets = spreadsheet.getSheets();
+
+    Logger.log("Preparing spreadsheet");
+    try {
+        allExtensionsSheet = spreadsheet.getSheetByName("All Extensions").clear();
+    } catch (e) {
+        allExtensionsSheet = spreadsheet.insertSheet("All Extensions");
+    }
+
+    var extensionTabNames = Object.keys(extensionPlaceholderTypeIdToFieldIdsObjectMapping).map(
+        function (key) { return extensionPlaceholderTypeIdToFieldIdsObjectMapping[key].name }
+    )
+    for (var i = 0, n = sheets.length; i < n; i++) {
+        if (extensionTabNames.indexOf(sheets[i].getName()) !== -1) {
+            spreadsheet.deleteSheet(sheets[i]);
+        }
+    }
+
+    // Get report
+    Logger.log("Getting report");
+    if (typeof includeExtensionsWithZeroImpressions === 'undefined') {
+        Logger.log("Warning: options variable includeExtensionsWithZeroImpressions not found. " +
+            "Continuing, and by default including extensions that have had no impressions today."
+        );
+        impressionsCondition = '';
+    } else {
+        if (includeExtensionsWithZeroImpressions === true) {
+            var impressionsCondition = '';
+        } else if (includeExtensionsWithZeroImpressions === false) {
+            var impressionsCondition = 'WHERE Impressions > 0 ';
+        } else {
+            throw ("Value of options variable includeExtensionsWithZeroImpressions must be a boolean, either true or false. " +
+                "Currently, it is: " + includeExtensionsWithZeroImpressions)
+        }
+    }
+
+    var placeholderFeedItemReport = AdsApp.report(
+        'SELECT PlaceholderType, AttributeValues, DisapprovalShortNames ' +
+        'FROM PLACEHOLDER_FEED_ITEM_REPORT ' +
+        impressionsCondition +
+        'DURING TODAY'
+    );
+    var reportRows = placeholderFeedItemReport.rows();
+
+    // Process report data, to prepare for outputting to the spreadsheet
+    Logger.log("Processing report data");
+    var extensionPlaceholderTypeIds = Object.keys(extensionPlaceholderTypeIdToFieldIdsObjectMapping);
+
+    var extensionTypeFeedItemCounts = {};
+    for (var i = 0, n = extensionPlaceholderTypeIds.length; i < n; i++) {
+        extensionTypeFeedItemCounts[extensionPlaceholderTypeIds[i]] = 0;
+    }
+    var jsonAttributeValuesByExtensionType = {};
+    var disapprovalShortNamesByExtensionType = {};
+    var duplicateAttributeValuesCounts = {};
+
+    while (reportRows.hasNext()) {
+        var row = reportRows.next();
+
+        var placeholderTypeId = row['PlaceholderType']
+        if (extensionPlaceholderTypeIds.indexOf(placeholderTypeId) === -1) {
+            continue;
+        }
+        if (Object.keys(jsonAttributeValuesByExtensionType).indexOf(placeholderTypeId) === -1) {
```

Comment:
To check if a javascript object has a key you can do `if("mykey" in myObject)` instead of using index of - might look nicer? Up to you which you prefer

Reply:
Oh yes, that's nicer!